File: src/Samples/Moq.VisualBasic/RefOut.vb
Imports System.Runtime.InteropServices
Imports Moq

Namespace Test

    Public Class Foo

        Public Sub Test()
            Dim parser = Mock.Of(Of IParser)

            parser.Setup(Of TryParse)(AddressOf parser.TryParse) _
                .Returns(Function(input As String, ByRef result As DateTimeOffset) DateTimeOffset.TryParse(input, result))

            Dim expected = DateTimeOffset.Now
            Dim value = expected.ToString("O")
            Dim actual As DateTimeOffset

            Debug.Assert(parser.TryParse(value, actual))
            Debug.Assert(actual = expected)

        End Sub

        Public Sub Test2()
            Dim target = Mock.Of(Of IEnvironment)

            Dim expected = DateTimeOffset.Now
            Dim value = expected.ToString("O")
            Dim actual As DateTimeOffset

            target.Setup(Of TryParse)(Function() AddressOf target.Parser.TryParse) _
                .Returns(Function(input As String, ByRef result As DateTimeOffset) DateTimeOffset.TryParse(input, result))

            Debug.Assert(target.Parser.TryParse(value, actual))
            Debug.Assert(actual = expected)

        End Sub

        Delegate Function TryParse(input As String, ByRef result As DateTimeOffset) As Boolean
    End Class

End Namespace

File: src/Samples/Moq.CSharp/RefOut.cs
using System;
using System.Diagnostics;
using Moq;
using Moq.CSharp;

namespace Sample.CSharp
{
    public class Foo
    {
        public void Test()
        {
            var mock = Mock.Of<IParser>();

            mock.Setup<TryParse>(mock.TryParse)
                .Returns((string input, out DateTimeOffset date) => DateTimeOffset.TryParse(input, out date));

            var expected = DateTimeOffset.Now;
            var value = expected.ToString("O");

            Debug.Assert(mock.TryParse(value, out var actual));
            Debug.Assert(actual == expected);
        }

        public void Test2()
        {
            var mock = Mock.Of<IRefOutParent>();

            var expected = DateTimeOffset.Now;
            var value = expected.ToString("O");

            mock.Setup<TryParse>(() => mock.RefOut.TryParse)
                .Returns((string input, out DateTimeOffset date) => DateTimeOffset.TryParse(value, out date));

            Debug.Assert(mock.RefOut.TryParse(value, out var actual));
            Debug.Assert(actual == expected);
        }

        delegate bool TryParse(string input, out DateTimeOffset date);
    }

    public interface IRefOutParent
    {
        IRefOut RefOut { get; }
    }

    public interface IRefOut
    {
        bool TryParse(string input, out DateTimeOffset date);
    }
}